### PR TITLE
Enables np.sum for timedelta64

### LINKED ISSUE #4540: Problem lowering code using timedelta arrays
Certain numba implementations of numpy functions rely on the `type` function. For example: 

https://github.com/numba/numba/blob/60d2bddd0a9ddf84cf900a70e4b7e17bcbe5dc87/numba/targets/arraymath.py#L246

However, `type` does not work with `timedelta64`, which means that for example, `np.sum(array, axis)` does not work with `timedelta64`.

```python
@numba.njit
def foo(x):
    return np.sum(x,0)

foo(np.ones((2,2))) #works
baz(np.ones((2,2)).astype(np.timedelta64)) #fails
```

This comes from the failure to run `type` on `timedelta64`

```python
def foo(x):
    return 1


@numba.extending.overload(foo)
def foo_fc(x):
    zero = getattr(x, 'dtype', x)(0)
    def foo_impl(x):
        a = type(zero)
        return x
    return foo_impl
    
@numba.njit
def bar(x):
    return foo(x)

bar(np.ones(2)) #works
bar(np.ones(2).astype(np.timedelta64)) #fails
```

### PR DESCRIPTION
**Pull request by luk-f-a**
fixes #4540

enables `type(timedelta64)` and therefore `np.sum` on `timedelta64` arrays. Adds tests for this case.

### CODE REVIEW

**Comment by luk-f-a** [comment]
ping @stuartarchibald 

`test_sum_axis_kws1` in numba/tests/test_array_methods.py should pass but returns

```
SubTest error: Traceback (most recent call last):
  File "/home//anaconda3/envs/numbaenv/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/home//anaconda3/envs/numbaenv/lib/python3.6/unittest/case.py", line 523, in subTest
    yield
  File "/home//mypyprojects/numba/numba/tests/test_array_methods.py", line 1019, in test_sum_axis_kws1
    cfunc(arr, axis=axis))
  File "/home//mypyprojects/numba/numba/dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home//mypyprojects/numba/numba/dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "/home//mypyprojects/numba/numba/six.py", line 668, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function gen_sum_axis_impl.<locals>.inner at 0x7f35d617e730>) with argument(s) of type(s): (array(timedelta64[M], 1d, C), int64)
 * parameterized
In definition 0:
    LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
unsupported type for input operand: timedelta64[M]

File "targets/arraymath.py", line 225:
    def inner(arr, axis):
        <source elided>
                                                       const_axis_val)
                result += arr[index_tuple_generic]
                ^

[1] During: lowering "$128.11 = inplace_binop(fn=<built-in function iadd>, immutable_fn=<built-in function add>, lhs=result, rhs=$128.10, static_lhs=Undefined, static_rhs=Undefined)" at /home//mypyprojects/numba/numba/targets/arraymath.py (225)
    raised from /home//mypyprojects/numba/numba/six.py:669
In definition 1:
    LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
unsupported type for input operand: timedelta64[M]

File "targets/arraymath.py", line 225:
    def inner(arr, axis):
        <source elided>
                                                       const_axis_val)
                result += arr[index_tuple_generic]
                ^

[1] During: lowering "$128.11 = inplace_binop(fn=<built-in function iadd>, immutable_fn=<built-in function add>, lhs=result, rhs=$128.10, static_lhs=Undefined, static_rhs=Undefined)" at /home//mypyprojects/numba/numba/targets/arraymath.py (225)
    raised from /home//mypyprojects/numba/numba/six.py:669
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function gen_sum_axis_impl.<locals>.inner at 0x7f35d617e730>)
[2] During: typing of call at /home//mypyprojects/numba/numba/targets/arraymath.py (342)


File "targets/arraymath.py", line 342:
    def array_sum_impl_axis(arr, axis):
        return compiled(arr, axis)
        ^

[1] During: lowering "$0.5 = call $0.2(func=$0.2, args=[], kws=[('axis', Var(axis, test_array_methods.py:189))], vararg=None)" at /home//mypyprojects/numba/numba/tests/test_array_methods.py (189)


One or more subtests failed
Failed subtests list: numba.tests.test_array_methods.TestArrayMethods.test_sum_axis_kws1 [Testing np.sum(axis) with timedelta64[M] input ] (test sum with axis parameter over a whole range of dtypes)


Ran 1 test in 29.222s

FAILED (errors=1)
```

**Comment by stuartarchibald** [comment]
This is a small reproducer of the problem:
```python
from numba import njit
import numpy as np
import logging

@njit(debug=True)
def FOO(arr, acc):
    arr + acc

x = np.ones((5,3), dtype='timedelta64[M]')
acc = np.zeros(x.shape, dtype=x.dtype)

FOO(x, acc)
```
after much digging around in the ufunc rewrite code, I believe the issue is here:
https://github.com/numba/numba/blob/0039fd3f17c24f5f8abae7171cca91162444d0f3/numba/typing/npydecl.py#L131-L133
The `ret_tys` comes out as:
```
UFuncLoopSpec(inputs=[timedelta64[M], timedelta64[M]], outputs=[timedelta64[]], ufunc_sig='mm->m')
```
which is clearly going to cause problems.

**Comment by luk-f-a** [comment]
@stuartarchibald , it looks like a full solution might take some time (I guess that's why you tagged it "0.50"). If that's the case, I propose that we proceed with this PR which adds support for `timedelta64` N-D arrays (for N>1) and that I open a new issue for the error with 1-d arrays. 
Even if not everything is working it's still an improvement from the current situation.
What do you think?

**Comment by luk-f-a** [comment]
@stuartarchibald : I found what the problems with `timedelta64' were. It's ready for review. After this PR is approved, #4480 can follow shortly afterwards.

**Comment by luk-f-a** [comment]
@stuartarchibald this is actually ready-to-review, in case you want to update the label.

**Comment by luk-f-a** [comment]
Regarding the question, what situation does the following code cover: https://github.com/numba/numba/pull/5060/files#diff-a5ddf7c956a12ebde493a9d9dc17e295R386-R399.

this attempts to address the issue  identified in https://github.com/numba/numba/pull/5060#issuecomment-573148173

In the current implementation, running the example in the comment leads to 
`ret_tys = [timedelta64[]]`
and
 `ufunc_loop = UFuncLoopSpec(inputs=[timedelta64[M], timedelta64[M]], outputs=[timedelta64[]], ufunc_sig='mm->m')`

which eventually leads to the error
`No conversion from timedelta64[M] to timedelta64[] for '$0.4'`

After the changes in the PR, 
`ret_tys = [timedelta64[M]]`
 `ufunc_loop = UFuncLoopSpec(inputs=[timedelta64[M], timedelta64[M]], outputs=[timedelta64[M]], ufunc_sig='mm->m')`
which prevents the converstion problem.

The code https://github.com/numba/numba/pull/5060/files#diff-a5ddf7c956a12ebde493a9d9dc17e295R386-R399 takes care of cases where we have a `timedelta[]` case (ie the output time unit is undefined) and tries to define the units based on the input units. It checks the units of the inputs, and if they all have the same units, it will pass that unit to the output (only if the output is also a timedelta).

I can add that explanation in the docstring if it helps.

File: numba/tests/test_array_methods.py [hunk 1]
```diff
@@ -1092,7 +1086,7 @@ def test_sum_axis_dtype_kws(self):
                       np.dtype('uint64'): [np.float64, np.uint64],
                       np.dtype('complex64'): [np.complex64, np.complex128],
                       np.dtype('complex128'): [np.complex128],
-                      np.dtype('timedelta64'): [np.timedelta64]}
+                      np.dtype('timedelta64[M]'): [np.dtype('timedelta64[M]')]}
 
         for arr_list in all_test_arrays:
             for arr in arr_list:
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py:1092]: Perhaps also tests a `Y` unit or something?

File: numba/np/numpy_support.py [hunk 2]
```diff
@@ -382,6 +382,21 @@ def choose_types(numba_types, ufunc_letters):
                   for letter in ufunc_letters[len(numba_types):]]
         return types
 
+    def set_output_dt_units(inputs, outputs):
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/np/numpy_support.py:385]: done

File: numba/tests/test_array_methods.py [hunk 3]
```diff
@@ -980,15 +980,12 @@ def test_sum_axis_kws1(self):
         pyfunc = array_sum_axis_kws
         cfunc = jit(nopython=True)(pyfunc)
         all_dtypes = [np.float64, np.float32, np.int64, np.uint64, np.complex64,
-                      np.complex128]
-        # timedelta test cannot be enabled until issue #4540 is fixed
-        # all_dtypes += [np.timedelta64]
+                      np.complex128, 'timedelta64[M]']
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/tests/test_array_methods.py:983]: done

File: numba/tests/test_array_methods.py (hunk 1 above)
Comment (luk-f-a) [reply numba/tests/test_array_methods.py:1092]: added

File: numba/np/numpy_support.py [hunk 4]
```diff
@@ -383,6 +383,20 @@ def choose_types(numba_types, ufunc_letters):
...
+        """
+        Sets the output unit of a datetime type based on the input units
+
+        Timedelta is a special dtype that requires the time unit to be
+        specified (day, month, etc). Not every operation with timedelta inputs
+        leads to an output of timedelta output. However, for those that do,
+        the unit of output must be inferred based on the units of the inputs.
+
+        At the moment this function takes care of the case where all
+        inputs have the same unit, and therefore the same as the output unit.
```
Comment (stuartarchibald) [inline numba/np/numpy_support.py:395]:
```suggestion
        inputs have the same unit, and therefore the output unit has the same.
```

File: numba/np/numpy_support.py [hunk 5]
```diff
@@ -383,6 +383,20 @@ def choose_types(numba_types, ufunc_letters):
...
+        If in the future this should be extended to a case with mixed units,
+        the rules should be implemented `npdatetime_helpers` and called from
```
Comment (stuartarchibald) [inline numba/np/numpy_support.py:397]:
```suggestion
        the rules should be implemented in `npdatetime_helpers` and called
```

File: numba/np/numpy_support.py [hunk 6]
```diff
@@ -383,6 +383,20 @@ def choose_types(numba_types, ufunc_letters):
...
+        this function to set the correct output unit.
```
Comment (stuartarchibald) [inline numba/np/numpy_support.py:398]:
```suggestion
        from this function to set the correct output unit.
```

**Pull request merged**